File: src/audit/validate.ts
/**
 * Author: Deepankar Das
 */

/**
 * Enforcer — Minimum Schema Validation Gate
 *
 * Every audit event must pass this gate before acceptance into the store.
 * Rejects events missing any of the 6 minimum gate fields.
 *
 * TDD Reference: Section 9.2
 * PRD Reference: Appendix C, R-3
 */

import { MINIMUM_GATE_FIELDS, type AuditEvent } from "../../types/audit-event.js";

export interface ValidationResult {
  valid: boolean;
  errors: string[];
}

/**
 * Validates that an audit event contains all minimum schema gate fields.
 * Each field must be present and non-empty.
 *
 * Gate fields: who, what, when, policy, decision, result
 */
export function validateAuditEvent(event: unknown): ValidationResult {
  const errors: string[] = [];

  if (event === null || event === undefined || typeof event !== "object") {
    return { valid: false, errors: ["Event is null, undefined, or not an object"] };
  }

  const record = event as Record<string, unknown>;

  for (const field of MINIMUM_GATE_FIELDS) {
    const value = record[field];
    if (value === undefined || value === null) {
      errors.push(`Missing required field: ${field}`);
    } else if (typeof value !== "string") {
      errors.push(`Field ${field} must be a string, got ${typeof value}`);
    } else if (value.trim().length === 0) {
      errors.push(`Field ${field} is empty`);
    }
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Builds the minimum gate fields from a full AuditEvent.
 * Used when constructing events to ensure gate fields are populated.
 */
export function buildGateFields(event: {
  actor: { user_id: string; agent_type: string };
  session_id: string;
  action: { type: string; attempted_action: string };
  timestamp: string;
  policy_detail: { policy_id: string; policy_version: string };
  decision: string;
  result: string;
}): Pick<AuditEvent, "who" | "what" | "when" | "policy" | "decision" | "result"> {
  return {
    who: `${event.actor.user_id}|${event.actor.agent_type}|${event.session_id}`,
    what: `${event.action.type}:${event.action.attempted_action}`,
    when: event.timestamp,
    policy: `${event.policy_detail.policy_id}@${event.policy_detail.policy_version}`,
    decision: event.decision,
    result: event.result,
  };
}

File: go/internal/audit/validate.go
/**
 * Author: Deepankar Das
 */

package audit

import (
	"fmt"

	"github.com/anthropics/enforcer/internal/types"
)

// ValidateAuditEvent checks the 6 minimum gate fields (who, what, when,
// policy, decision, result) are non-empty strings. Returns true with an
// empty slice if valid, or false with a list of missing field names.
func ValidateAuditEvent(event types.AuditEvent) (bool, []string) {
	var missing []string

	fields := map[string]string{
		"who":      event.Who,
		"what":     event.What,
		"when":     event.When,
		"policy":   event.Policy,
		"decision": event.Decision,
		"result":   event.Result,
	}

	for _, name := range types.MinimumGateFields {
		if fields[name] == "" {
			missing = append(missing, name)
		}
	}

	return len(missing) == 0, missing
}

// BuildGateFields constructs the 6 minimum gate fields from structured input.
// The "who" field is a compound of actor user_id, agent_type, and session_id.
// The "what" field combines action type and attempted_action.
// The "when" field is the timestamp.
// The "policy" field combines policy_id and policy_version.
// The "decision" and "result" fields are passed through directly.
func BuildGateFields(
	actor types.Actor,
	sessionID string,
	action types.ActionDetail,
	timestamp string,
	policyID string,
	policyVersion string,
	decision string,
	result string,
) map[string]string {
	who := fmt.Sprintf("%s|%s|%s", actor.UserID, actor.AgentType, sessionID)
	what := fmt.Sprintf("%s|%s", action.Type, action.AttemptedAction)
	policy := fmt.Sprintf("%s@%s", policyID, policyVersion)

	return map[string]string{
		"who":      who,
		"what":     what,
		"when":     timestamp,
		"policy":   policy,
		"decision": decision,
		"result":   result,
	}
}
